### PR TITLE
add support for the updateSecretDesc restful API

### PR DESCRIPTION
**Pull request by Yidingluo1**
add support for the updateSecretDesc restful API
this restful API will be exposed to public as:
POST <ehsm_srv_address>/ehsm?Action=UpdateSecretDesc
Interface is used to update the description of secret.
By entering a new description in payload, the description of the secretname entered in payload will be updated
add the corresponding test cases.
test: passed the unittest

Signed-off-by: yidingluo1 <luo.yd@neusoft.com>

### CODE REVIEW

File: ehsm_kms_service/router.js [hunk 1]
```diff
@@ -25,6 +25,7 @@ const {
 } = require('./key_management_apis')
 const {
   createSecret,
+  updateSecret
```
Comment (syan10) [inline ehsm_kms_service/router.js:28]: so why rename the api to updatesecret?

File: ehsm_kms_service/secret_manager_apis.js [hunk 2]
```diff
@@ -234,9 +234,68 @@ const createSecret = async (res, appid, payload, DB) => {
     }
 }
 
+//updateSecret
```
Comment (syan10) [inline ehsm_kms_service/secret_manager_apis.js:237]: add more detail description for the API

File: ehsm_kms_service/secret_manager_apis.js [hunk 3]
```diff
@@ -234,9 +234,68 @@ const createSecret = async (res, appid, payload, DB) => {
...
+            'nextRotationDate'
+        ],
+        limit: 1,
+    }
+    await DB.partitionedFind('secret_metadata', secret_name_query)
+        .then(async (secret_metadata_res) => {
+            if (secret_metadata_res.docs.length > 0) {
+                secret_metadata_res.docs[0].description = description
+                DB.insert(secret_metadata_res.docs[0])
+                    .then(() => {
```
Comment (syan10) [inline ehsm_kms_service/secret_manager_apis.js:278]: how about the version handling? it should be automatically increased right?

File: ehsm_kms_service/secret_manager_apis.js (hunk 2 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:237]: it will be add more description

File: ehsm_kms_service/secret_manager_apis.js (hunk 3 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:278]: The version will increase automatically. This function will be completed in the put secret value interface

File: ehsm_kms_service/router.js (hunk 1 above)
Comment (wanghouqi) [reply ehsm_kms_service/router.js:28]: The name will be change to UpdateSecretDesc

**Comment by syan10** [comment]
same with the pervious commit, you need explain more details, such as this API's purpose?

**Comment by Yidingluo1** [comment]
> same with the pervious commit, you need explain more details, such as this API's purpose?

commit changed

File: ehsm_kms_service/secret_manager_apis.js [hunk 4]
```diff
@@ -234,9 +234,68 @@ const createSecret = async (res, appid, payload, DB) => {
...
+            'deleteTime',
+            'plannedDeleteTime',
+            'rotationInterval',
+            'lastRotationDate',
+            'nextRotationDate'
+        ],
+        limit: 1,
+    }
+    await DB.partitionedFind('secret_metadata', secret_name_query)
+        .then(async (secret_metadata_res) => {
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:274]: The async used here, makes this update multi-thread?
How can we keep the data‘s consistency while using multi-thread? Do we have any lock if a put and a update comes in at the same time?

File: ehsm_kms_service/secret_manager_apis.js [hunk 5]
```diff
@@ -234,9 +234,68 @@ const createSecret = async (res, appid, payload, DB) => {
...
+            if (secret_metadata_res.docs.length > 0) {
```
Comment (HawieLong) [inline ehsm_kms_service/secret_manager_apis.js:275]: If I want to update my description from "des" to "". Will this code block this?

File: test/cli/updateSecretDesc.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    parser.add_argument('--secretName', type=str, help='the name of the secret', required=True)
+    parser.add_argument('--description', type=str, help='the description of the secret', required=True)
+    args = parser.parse_args()
+
+    base_url = args.url + "/ehsm?Action="
+    print(base_url)
+    return base_url, args.secretName, args.description
+
+def updateSecretDesc(base_url, secretName, description):
+    print('update secret')
```
Comment (HawieLong) [inline test/cli/updateSecretDesc.py:27]: update secret description

Comment (Yidingluo1) [reply test/cli/updateSecretDesc.py:27]: it will be changed

File: ehsm_kms_service/secret_manager_apis.js (hunk 4 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:274]: async has no effect here and will be deleted

File: ehsm_kms_service/secret_manager_apis.js (hunk 5 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:275]: ""is not supported，it will sign error

Comment (HawieLong) [reply ehsm_kms_service/secret_manager_apis.js:275]: payload["description"] = ""
why sign error?

Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:275]: yes

File: ehsm_kms_service/secret_manager_apis.js (hunk 4 above)
Comment (HawieLong) [reply ehsm_kms_service/secret_manager_apis.js:274]: OK, the database lock issue can be fixed after we discuss it, maybe in other patches.

File: ehsm_kms_service/secret_manager_apis.js (hunk 5 above)
Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:275]: ”“ in any payload will cause signerror

Comment (HawieLong) [reply ehsm_kms_service/secret_manager_apis.js:275]: That should be a sign problem. We should fix this and support "". Will discuss it later.

Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:275]: Got it

File: ehsm_kms_service/secret_manager_apis.js [hunk 7]
```diff
@@ -234,9 +234,69 @@ const createSecret = async (res, appid, payload, DB) => {
...
+            'rotationInterval',
+            'lastRotationDate',
+            'nextRotationDate'
+        ],
+        limit: 1,
+    }
+    await DB.partitionedFind('secret_metadata', secret_name_query)
+        .then((secret_metadata_res) => {
+            if (secret_metadata_res.docs.length > 0) {
+                secret_metadata_res.docs[0].description = description
```
Comment (yang8621) [inline ehsm_kms_service/secret_manager_apis.js:276]: Are there any rules to checking the length of the description?  How about a 4GB description inserted?

Comment (yang8621) [reply ehsm_kms_service/secret_manager_apis.js:276]: Can we truncate to the first 4K chars?

Comment (Yidingluo1) [reply ehsm_kms_service/secret_manager_apis.js:276]: The limit will be added: the length will not exceed 4*1024

**Comment by syan10** [comment]
new commit title
"
add support for the updateSecretDesc restful API

this restful API will be exposed to public as:
POST <ehsm_srv_address>/ehsm?Action=UpdateSecretDesc

which is used to update the description of secret.

test: passed the unittest
"

**Pull request closed without merging**